oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenIntrospectionEndpointFilter.java

```diff
@@ -0,0 +1,240 @@
+					.isTokenActive()
+							? generateTokenIntrospectionResponse(
+									introspectionTokenAuthentication.getTokenHolder(),
+									introspectionTokenAuthentication.getClientId())
+							: OAuth2TokenIntrospectionResponse.builder(false).build();
+
+			this.sendTokenIntrospectionResponse(response, tokenIntrospectionResponse);
+		} catch (OAuth2AuthenticationException ex) {
+			SecurityContextHolder.clearContext();
+			sendErrorResponse(response, ex.getError());
+		}
+	}
+
+	private OAuth2TokenIntrospectionResponse generateTokenIntrospectionResponse(
+			Token<? extends AbstractOAuth2Token> tokenHolder, String clientId) {
+		AbstractOAuth2Token token = tokenHolder.getToken();
+		OAuth2TokenIntrospectionResponse.Builder builder = OAuth2TokenIntrospectionResponse
+				.withClaims(tokenHolder.getClaims());
+		builder.active(true).clientId(clientId);
+		Optional.ofNullable(token.getIssuedAt()).ifPresent(builder::issuedAt);
+		Optional.ofNullable(token.getExpiresAt()).ifPresent(builder::expirationTime);
+		TokenToIntrospectionResponseFieldsMapper.extractFromToken(token, builder);
+		return builder.build();
+	}
+
+	private void sendErrorResponse(HttpServletResponse response, OAuth2Error error) throws IOException {
+		ServletServerHttpResponse httpResponse = new ServletServerHttpResponse(response);
+		httpResponse.setStatusCode(HttpStatus.BAD_REQUEST);
+		this.errorHttpResponseConverter.write(error, null, httpResponse);
+	}
+
+	private void sendTokenIntrospectionResponse(HttpServletResponse response,
+			OAuth2TokenIntrospectionResponse tokenIntrospectionResponse) throws IOException {
+		ServletServerHttpResponse httpResponse = new ServletServerHttpResponse(response);
+		this.tokenIntrospectionHttpResponseConverter.write(tokenIntrospectionResponse, null, httpResponse);
+	}
+
+	private static void throwError(String errorCode, String parameterName) {
+		OAuth2Error error = new OAuth2Error(
+				errorCode, "OAuth 2.0 Token Introspection Parameter: " + parameterName,
+				"https://tools.ietf.org/html/rfc7662#section-2.1");
+		throw new OAuth2AuthenticationException(error);
+	}
+
+	/**
+	 * Mapper that helps populate {@code OAuth2TokenIntrospectionResponse} fields from different {@code AbstractOAuth2Token}
+	 * implementations.
+	 *
+	 * @see OAuth2AccessToken
+	 *
+	 * @author Gerardo Roza
+	 */
+	private static final class TokenToIntrospectionResponseFieldsMapper {
+
+		private static final Map<Class<? extends AbstractOAuth2Token>, BiConsumer<AbstractOAuth2Token, OAuth2TokenIntrospectionResponse.Builder>> supportedTokens;
+		static {
+			Map<Class<? extends AbstractOAuth2Token>, BiConsumer<AbstractOAuth2Token, OAuth2TokenIntrospectionResponse.Builder>> tokenMap = new HashMap<>();
+			tokenMap.put(
+					OAuth2AccessToken.class,
+					(token, builder) -> extractFromOAuth2AccessToken((OAuth2AccessToken) token, builder));
+			supportedTokens = Collections.unmodifiableMap(tokenMap);
```

Comment:
I know this mapper doesn't make much sense now that we have only one entry (before it contained another entry for `Jwt`) but maybe we want to leave it as a flexible solution in case we can extract relevant information from Refresh Tokens or, more likely, from ID Tokens in the future.

Reply:
Let's remove this as it's not needed.

In `OAuth2TokenIntrospectionAuthenticationToken` remove `clientId` and `tokenHolder` members and add member `Map<String, Object> claims`. The `OAuth2TokenIntrospectionAuthenticationProvider` has all the context required to produce the `claims` that will be returned in the Token Introspection response via `OAuth2TokenIntrospectionEndpointFilter` as-is.

Reply:
Cool. My idea was to separate as much as possible the response representation model classes from the framework model classes, that is, in the provider handle objects that belong to the core of the domain, and then use the EndpointFilter to map that to the expected response (kind of a layer separation). Of course, the `claims` is actually a plain map, but the structure, the keys it contains are strictly related to the response that is finally serialized. 
Having clarified the reasons why I chose that approach (probably because I'm more used to enterprise solutions, some rules don't totally apply to frameworks/libraries), I understand the reasons for your suggestion, it makes sense. I now changes this (y)